### ISSUE TITLE
Service containers not being included in VIP pool

### ISSUE THREAD

**Issue by dariko** [docker/docker#25219]
**Output of `docker version`:**

```
$ docker version
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 21:42:26 2016
 OS/Arch:      linux/amd64
 Experimental: true

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 21:42:26 2016
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
$ docker info
Containers: 8
 Running: 0
 Paused: 0
 Stopped: 8
Images: 2926
Server Version: 1.12.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 2142
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host bridge overlay
Swarm: pending
 NodeID: 84r7yfiwd2w83sn6kt4dms32h
 Is Manager: false
 Node Address: 192.168.99.101
Runtimes: runc
Default Runtime: runc
Security Options:
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 3.692 GiB
Name: ball
ID: DFT5:E4J5:HBYF:E4C4:ALCK:Z46W:E3AJ:LQK5:L6FQ:QFKF:3EPD:2SPA
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No kernel memory limit support
WARNING: No oom kill disable support
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
Experimental: true
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

running on docker-machine using 1.12.0 image
$ docker-machine version
docker-machine version 0.8.0-rc2, build 4ca1b85

I noticed some containers with containers not responding trough the VIP
in swarm mode, so I wrote the following test script:
```
#!/bin/bash

CREATE_VMS=1
if [ ! -z CREATE_VMS ]
then
  docker-machine rm -f s1 s2
  docker-machine create -d virtualbox --virtualbox-cpu-count=2 --virtualbox-memory 1024 s1 &
  docker-machine create -d virtualbox --virtualbox-cpu-count=2 --virtualbox-memory 1024 s2 &
  wait
fi

docker $(docker-machine config s1) swarm leave --force
docker $(docker-machine config s2) swarm leave --force
# configure s1 as first swarm manager
docker $(docker-machine config s1) swarm init --advertise-addr $(docker-machine ip s1) --listen-addr $(docker-machine ip s1)
# read token
token=$(docker $(docker-machine config s1) swarm join-token -q manager)
# join s2 to swarm as manager
docker $(docker-machine config s2) swarm join \
    --advertise-addr $(docker-machine ip s2) \
    --listen-addr $(docker-machine ip s2) \
    --token "$token" \
    $(docker-machine ip s1)


function wait_for_running_container_count() {
  required_count=$1
  containers_count=$(docker $(docker-machine config s1) service ps hostname |awk '{ if($6=="Running") print $1}'|wc -l)
  while [ $containers_count -ne $required_count ]
  do
    echo "waiting for containers to be running, now $containers_count/$required_count"
    sleep 3
    containers_count=$(docker $(docker-machine config s1) service ps hostname |awk '{ if($6=="Running") print $1}'|wc -l)
  done
  echo "got $containers_count/$required_count containers running"
}

function wait_for_responding_container_count() {
  required_count=$1
  service_ip="$2"
  max_count=$3
  export service_ip
  containers_count=$(parallel --no-notice --env service_ip 'curl -s $service_ip' ::: $(seq 1 $(( $max_count * 2 )) ) |sort -u|wc -l)
  while [ $containers_count -ne $required_count ]
  do
    echo "waiting for correct number of containers to respond via $service_ip, got $containers_count/$required_count containers"
    sleep 3
    containers_count=$(parallel --no-notice --env service_ip 'curl -s $service_ip' ::: $(seq 1 $(( $max_count * 2 )) ) |sort -u|wc -l)
  done
  echo "correct number of container responding via ${service_ip}: $containers_count/$required_count"
}

docker $(docker-machine config s1) service create --replicas 2 --name hostname --publish 80:80 alpine \
  sh -c 'while true; do echo -e "HTTP/1.1 200 OK\n\n$(hostname)" | nc -lp 80; done'
wait_for_running_container_count 2
wait_for_responding_container_count 2 $(docker-machine ip s1) 5
wait_for_responding_container_count 2 $(docker-machine ip s2) 5

MAXCONT=30
while true
do
  new_scale=$(( $RANDOM % $MAXCONT +1 ))
  # scale container 
  echo "Scaling service hostname to $new_scale"
  docker $(docker-machine config s1) service scale hostname=$new_scale
  wait_for_running_container_count $new_scale
  sleep 1
  for service_ip in "$(docker-machine ip s1)" "$(docker-machine ip s2)"
  do
    wait_for_responding_container_count $new_scale "$service_ip" $MAXCONT
  done
  sleep 1
done
```

This script builds a swarm comprised of 2 managers and add a service with
a published port.
It should then keep scaling the service and checking if the service
has been scaled correctly both as numbers of containers started and as
number of containers receiving requests through the VIP.

In practice after some iteration of the loop the service stops responding through all the containers.
As an example my last run stopped after trying a `service scale=4 hostname`
```
got 4/4 containers running
waiting for correct number of containers to respond via 192.168.99.101, got 3/4 containers
waiting for correct number of containers to respond via 192.168.99.101, got 3/4 containers
waiting for correct number of containers to respond via 192.168.99.101, got 3/4 containers
waiting for correct number of containers to respond via 192.168.99.101, got 3/4 containers
waiting for correct number of containers to respond via 192.168.99.101, got 3/4 containers
waiting for correct number of containers to respond via 192.168.99.101, got 3/4 containers
waiting for correct number of containers to respond via 192.168.99.101, got 3/4 containers
waiting for correct number of containers to respond via 192.168.99.101, got 3/4 containers
waiting for correct number of containers to respond via 192.168.99.101, got 3/4 containers
waiting for correct number of containers to respond via 192.168.99.101, got 3/4 containers
waiting for correct number of containers to respond via 192.168.99.101, got 3/4 containers
waiting for correct number of containers to respond via 192.168.99.101, got 3/4 containers
```

manually verifying the situation I noticed there are in fact 4 containers running
```
$ docker $(docker-machine config s2) service ps hostname|tail -n +2|sort -k 6
0i6gnk22s4bpfz4qvuxk5gpbz  hostname.9       alpine  s1    Running        Running 6 minutes ago   
0sqty0sjr77kf4xwq9xh8cpyu  hostname.6       alpine  s1    Running        Running 6 minutes ago   
127kjdk6r0rl6gcv2skiyku1s  hostname.16      alpine  s2    Running        Running 6 minutes ago   
149tkg6p51ffz2p9oa4kbq6v6  hostname.2       alpine  s2    Running        Running 7 minutes ago   
32ib60df3bpkdkvxa7gkdfqoe  hostname.1       alpine  s1    Shutdown       Shutdown 5 minutes ago  
3djpi5b5koqthi30zc9xji5at  hostname.8       alpine  s1    Shutdown       Shutdown 5 minutes ago  
4hgx2dofyaw89k3dq65acvswh  hostname.14      alpine  s1    Shutdown       Shutdown 5 minutes ago  
[ some other state=desired-state=shutdown containers ]
```

but requests to each of the docker-machine ips is routed only to 3 of them
```
$ ip=$(docker-machine ip s1);for i in $(seq 1 30);do curl -s $ip;done|sort|uniq
75112f2347a2
8a68ab140779
d525c7ba38aa
$ ip=$(docker-machine ip s2);for i in $(seq 1 30);do curl -s $ip;done|sort|uniq
75112f2347a2
8a68ab140779
d525c7ba38aa
```

The running containers are
```
$ for machine_name in s1 s2;do echo $machine_name; docker $(docker-machine config $machine_name) ps -q;done
s1
75112f2347a2
8a68ab140779
s2
2bfe255dfe65
d525c7ba38aa
```

So the container not reached by the requests to the VIPs is  2bfe255dfe65.
This container has been assigned 10.255.0.12 as IP on the ingress network
```
$ docker $(docker-machine config s2) inspect 2bfe255dfe65|jq '.[0].NetworkSettings.Networks.ingress.IPAddress'
"10.255.0.12"
```
which seems to be reachable from containers in both docker-machines:
```
docker $(docker-machine config s1) exec -it 75112f2347a2 sh -c 'echo | nc 10.255.0.12 80'
HTTP/1.1 200 OK

2bfe255dfe65
$ docker $(docker-machine config s2) exec -it d525c7ba38aa sh -c 'echo | nc 10.255.0.12 80'
HTTP/1.1 200 OK

2bfe255dfe65
```

In both the s1,s2:/var/log/docker.log I  found various occurrences of the errors
* `could not resolve peer \"10.255.0.3\": could not resolve peer: serf instance not initialized`
* `Failed to delete real server 10.255.0.21 for vip 10.255.0.5 fwmark 256: no such file or directory`
* `container status unavailable" error="context canceled" module=taskmanager task.id=5r8bxgf2oq726yvpfcmjqfm8s`

I noticed the problem seems to appear always after a scaling down request, but I
haven't done enough testing to be confident on this.

I'm attaching the docker-machine's /var/log/docker.log related to the described
script run:
[s1-var_log_docker.txt](https://github.com/docker/docker/files/390296/s1-var_log_docker.txt)
[s2-var_log_docker.txt](https://github.com/docker/docker/files/390297/s2-var_log_docker.txt)

edited script: `--no-notice` flag for `parallel` and 1024 MB ram for each manager

**Comment by thaJeztah** [comment]
ping @mrjana PTAL

**Comment by jhorwit2** [comment]
I can't reproduce this. 
```
docker service create --replicas 3 --name swarmtest --network backend --publish 8085:80 jhorwit2/swarm-test:0.0.2
```

I ran curl on each swarm host and got RR between 3 ips.

Then i ran 

`docker service scale swarmtest=5`

Repeated the curl test and got RR between 5 different ips on each swarm host.

**Comment by dariko** [comment]
If I run it one time it works, the problems begins to appear after some scaling.
I just tried my script it again in ubuntu 1504 (adding a --no-notice flag to parallel) (kernel 4.2.0, docker 1.12.0), and the service correctly scaled to 2, 5, 4, 16, 14. Each of these scaling was considered OK after checking that the number of containers responding to the requests made to a manager was consistent with the `scale=x` requested.

then the script executed a `docker .... service scale hostname=19`, and it got stuck not getting responses with 19 unique hostnames.

I stopped the script and checked the current situation. There are 19 containers running:
```
$ docker $(docker-machine config s1) service ps hostname|grep Running
2x04ky2cqzyjftxdb4dmctume  hostname.1       alpine  s2    Running        Running 11 minutes ago   
6kb1lg6sdhecvaz0ucmouvyg1  hostname.2       alpine  s2    Running        Running 12 minutes ago   
51i2gcrmste3en0bmtfq2jors  hostname.3       alpine  s2    Running        Running 11 minutes ago   
8z8pegwpezlpjgqfuectyrc7h  hostname.4       alpine  s1    Running        Running 12 minutes ago   
39azxevszhk5sop2arnpvink6  hostname.5       alpine  s1    Running        Running 12 minutes ago   
9amfxywlio6bjz5ff0v135c9j  hostname.6       alpine  s2    Running        Running 11 minutes ago   
836pu4ec5o2ny78o15usgkl7t  hostname.7       alpine  s2    Running        Running 11 minutes ago   
28yvqfyj79cyjgafziy1s8ztx  hostname.8       alpine  s1    Running        Running 11 minutes ago   
17qevlxozeecouys7svczvxr3  hostname.9       alpine  s1    Running        Running 11 minutes ago   
2x9sqg2wr9wma83jwhwmy6yzk  hostname.10      alpine  s1    Running        Running 11 minutes ago   
1ukwkc1is36ntokc08o51xvkq  hostname.11      alpine  s1    Running        Running 11 minutes ago   
7vjy7vybxfpri7s9695ygwaz9  hostname.12      alpine  s2    Running        Running 11 minutes ago   
d1ungk04ux7z1klwdhfdxgj0y  hostname.13      alpine  s2    Running        Running 11 minutes ago   
7ih1esznojfvihtvpq9ysblzl  hostname.14      alpine  s2    Running        Running 11 minutes ago   
0lrar75b8rvdmbr0k9v2v4lvo  hostname.15      alpine  s1    Running        Running 11 minutes ago   
4vtx658qovpkymimcv03b6ya6  hostname.16      alpine  s1    Running        Running 11 minutes ago   
emc0q0ko3dddja6gw0r9jyzff  hostname.17      alpine  s2    Running        Running 11 minutes ago   
9b47aw9sqczvnz8lrz25e7r3d  hostname.18      alpine  s1    Running        Running 11 minutes ago   
8z3aeehn2fpje10ufatkl0my4  hostname.19      alpine  s1    Running        Running 11 minutes ago
$ (docker $(docker-machine config s1) ps|tail -n +2;docker $(docker-machine config s2) ps|tail -n +2)|sort
18af44c18e6e        alpine:latest       "sh -c 'while true; d"   16 minutes ago      Up 16 minutes                           hostname.11.1ukwkc1is36ntokc08o51xvkq
2b8fa06cbd84        alpine:latest       "sh -c 'while true; d"   17 minutes ago      Up 17 minutes                           hostname.8.28yvqfyj79cyjgafziy1s8ztx
37e380c8ed63        alpine:latest       "sh -c 'while true; d"   17 minutes ago      Up 17 minutes                           hostname.10.2x9sqg2wr9wma83jwhwmy6yzk
45b4cc6c6a3a        alpine:latest       "sh -c 'while true; d"   17 minutes ago      Up 17 minutes                           hostname.7.836pu4ec5o2ny78o15usgkl7t
4647ed0cfd15        alpine:latest       "sh -c 'while true; d"   17 minutes ago      Up 17 minutes                           hostname.4.8z8pegwpezlpjgqfuectyrc7h
5a9055b0b62d        alpine:latest       "sh -c 'while true; d"   17 minutes ago      Up 17 minutes                           hostname.6.9amfxywlio6bjz5ff0v135c9j
5aaa625881c6        alpine:latest       "sh -c 'while true; d"   17 minutes ago      Up 17 minutes                           hostname.5.39azxevszhk5sop2arnpvink6
74ea52b4c429        alpine:latest       "sh -c 'while true; d"   17 minutes ago      Up 17 minutes                           hostname.12.7vjy7vybxfpri7s9695ygwaz9
858d731c71ec        alpine:latest       "sh -c 'while true; d"   17 minutes ago      Up 17 minutes                           hostname.1.2x04ky2cqzyjftxdb4dmctume
98c0b121b5be        alpine:latest       "sh -c 'while true; d"   16 minutes ago      Up 16 minutes                           hostname.3.51i2gcrmste3en0bmtfq2jors
9fbd6c769e2e        alpine:latest       "sh -c 'while true; d"   17 minutes ago      Up 17 minutes                           hostname.13.d1ungk04ux7z1klwdhfdxgj0y
a23907b5e292        alpine:latest       "sh -c 'while true; d"   16 minutes ago      Up 16 minutes                           hostname.17.emc0q0ko3dddja6gw0r9jyzff
a2e2fef19af6        alpine:latest       "sh -c 'while true; d"   17 minutes ago      Up 17 minutes                           hostname.9.17qevlxozeecouys7svczvxr3
a4382aa37557        alpine:latest       "sh -c 'while true; d"   17 minutes ago      Up 17 minutes                           hostname.14.7ih1esznojfvihtvpq9ysblzl
a62eb7f162b2        alpine:latest       "sh -c 'while true; d"   17 minutes ago      Up 17 minutes                           hostname.15.0lrar75b8rvdmbr0k9v2v4lvo
cb6837156382        alpine:latest       "sh -c 'while true; d"   16 minutes ago      Up 16 minutes                           hostname.18.9b47aw9sqczvnz8lrz25e7r3d
e042337bd1a9        alpine:latest       "sh -c 'while true; d"   17 minutes ago      Up 17 minutes                           hostname.16.4vtx658qovpkymimcv03b6ya6
e34705823fc4        alpine:latest       "sh -c 'while true; d"   16 minutes ago      Up 16 minutes                           hostname.19.8z3aeehn2fpje10ufatkl0my4
ff154aed0647        alpine:latest       "sh -c 'while true; d"   17 minutes ago      Up 17 minutes                           hostname.2.6kb1lg6sdhecvaz0ucmouvyg1
```
but requests to each of the 2 manager IPs gets routed to only 16 containers:
```
$ for i in $(seq 1 90);do curl -s 192.168.99.100;done|sort|uniq |sort
2b8fa06cbd84
37e380c8ed63
4647ed0cfd15
5a9055b0b62d
5aaa625881c6
74ea52b4c429
858d731c71ec
98c0b121b5be
9fbd6c769e2e
a2e2fef19af6
a4382aa37557
a62eb7f162b2
cb6837156382
e042337bd1a9
e34705823fc4
ff154aed0647
user@stick:~/src/docker_bug_25219$ for i in $(seq 1 90);do curl -s 192.168.99.101;done|sort|uniq |sort
2b8fa06cbd84
37e380c8ed63
4647ed0cfd15
5a9055b0b62d
5aaa625881c6
74ea52b4c429
858d731c71ec
98c0b121b5be
9fbd6c769e2e
a2e2fef19af6
a4382aa37557
a62eb7f162b2
cb6837156382
e042337bd1a9
e34705823fc4
ff154aed0647
```

Comparing those lists it appears that the containers with ID 18af44c18e6e (running on s1), and 45b4cc6c6a3a/a23907b5e292 (running on s2) are not partecipating in the responses. 
I then got the container addresses:
```
$ docker $(docker-machine config s1) inspect 18af44c18e6e  |jq '.[0].NetworkSettings.Networks.ingress.IPAddress'
"10.255.0.8"
$ docker $(docker-machine config s2) inspect 45b4cc6c6a3a  |jq '.[0].NetworkSettings.Networks.ingress.IPAddress'
"10.255.0.6"
$ docker $(docker-machine config s2) inspect a23907b5e292  |jq '.[0].NetworkSettings.Networks.ingress.IPAddress'
"10.255.0.12"
```

and tried to contact the containers without going through the virtual IP, but directly from sibling containers on the same network (even belonging to the same service). I picked e34705823fc4 on host s1 and 45b4cc6c6a3a on s2:
```
$ docker $(docker-machine config s1) exec -it e34705823fc4 sh -c "echo |nc 10.255.0.6 80"
HTTP/1.1 200 OK

45b4cc6c6a3a
$ docker $(docker-machine config s1) exec -it e34705823fc4 sh -c "echo |nc 10.255.0.8 80"
HTTP/1.1 200 OK

18af44c18e6e
$ docker $(docker-machine config s1) exec -it e34705823fc4 sh -c "echo |nc 10.255.0.12 80"
HTTP/1.1 200 OK

a23907b5e292
$ docker $(docker-machine config s2) exec -it 45b4cc6c6a3a sh -c "echo |nc 10.255.0.6 80"
HTTP/1.1 200 OK

45b4cc6c6a3a
$ docker $(docker-machine config s2) exec -it 45b4cc6c6a3a sh -c "echo |nc 10.255.0.8 80"
HTTP/1.1 200 OK

18af44c18e6e
$ docker $(docker-machine config s2) exec -it 45b4cc6c6a3a sh -c "echo |nc 10.255.0.12 80"
HTTP/1.1 200 OK

a23907b5e292
```


To recap: now my swarm is reporting the `hostname` service as having `scale=19`, and indeed there are 19 container tasks running in the swarm which are correctly working.
The problem is that the requests made to each manager IP gets routed only to 16 of these containers.

----
note: I just modified the script above, giving more memory to each manager (just to be sure:D) and adding a --no-notice flag to the parallel command which is needed in ubuntu 1504 while it wasn't on debian 8)

**Comment by mrjana** [comment]
@dariko Appreciate the detailed bug report and the scripting to repro the problem. I will try it on my end but it seems we are hitting some corner case scenario where the backend information for the LB is not getting properly updated.

**Comment by jhorwit2** [comment]
really strange. I ran the same scripts again scaling from 2 -> 10 -> 20 -> 15 and each time got all the ids. 

`docker service create --replicas 2 --name hostname --publish 80:80 alpine   sh -c 'while true; do echo -e "HTTP/1.1 200 OK\n\n$(hostname)" | nc -lp 80; done'`

`docker scale service hostname=10`

```
$ test h1 | sort | uniq | sort
066ff5517f6b
30a7b763aa40
4c125d4b000e
660cacc9bf2f
677e491bdf33
711fa5f2d43a
8bb47118dd3b
b5b244771109
e290b10b656e
e6e26dbe32aa

$ test h1 | sort | uniq | sort
066ff5517f6b
30a7b763aa40
4c125d4b000e
660cacc9bf2f
677e491bdf33
711fa5f2d43a
8bb47118dd3b
b5b244771109
e290b10b656e
e6e26dbe32aa
```

`docker service scale hostname=20`

```

$ test h1 | sort | uniq | sort
066ff5517f6b
0cc0d5388cb6
0e63d0752b04
30a7b763aa40
4c125d4b000e
5c262efd6160
6276ee721e3d
660cacc9bf2f
677e491bdf33
711fa5f2d43a
8bb47118dd3b
8d359d191fe0
aff6405047fa
b5b244771109
b6328e9456f6
cef409604850
d0d666193745
e290b10b656e
e6e26dbe32aa
f6fe2804a4fa

 test h1 | sort | uniq | sort
066ff5517f6b
0cc0d5388cb6
0e63d0752b04
30a7b763aa40
4c125d4b000e
5c262efd6160
6276ee721e3d
660cacc9bf2f
677e491bdf33
711fa5f2d43a
8bb47118dd3b
8d359d191fe0
aff6405047fa
b5b244771109
b6328e9456f6
cef409604850
d0d666193745
e290b10b656e
e6e26dbe32aa
f6fe2804a4fa
```

`docker service scale hostname=15`

```
test h1 | sort | uniq | sort
066ff5517f6b
0cc0d5388cb6
0e63d0752b04
4c125d4b000e
5c262efd6160
6276ee721e3d
660cacc9bf2f
711fa5f2d43a
8bb47118dd3b
aff6405047fa
b6328e9456f6
cef409604850
d0d666193745
e6e26dbe32aa
f6fe2804a4fa

test h2 | sort | uniq | sort
066ff5517f6b
0cc0d5388cb6
0e63d0752b04
4c125d4b000e
5c262efd6160
6276ee721e3d
660cacc9bf2f
711fa5f2d43a
8bb47118dd3b
aff6405047fa
b6328e9456f6
cef409604850
d0d666193745
e6e26dbe32aa
f6fe2804a4fa
```

**Comment by jhorwit2** [comment]
I wonder if this has something to do with the docker machine host? I don't have that setup to test. I'm running this on 3 ubuntu 15.04 vms.

**Comment by dariko** [comment]
@jhorwit2 are h1/h2 both managers? Later I'm gonna try it on 2 ubuntu 16.04 hosts.
I used docker-machine hoping for a more consistent setup. I was investigating another problem I'm having and I noticed this one :wink: 

@mrjana thank you

**Comment by jhorwit2** [comment]
@dariko yea, i have 3 managers only in that cluster.

**Comment by dariko** [comment]
Ok, I got to try it on two physical ubuntu 16.04, both on 1.12.0, exposing $ip1 and $ip3 on a private network (no firewall active).

Note that after each scaling request I waited until the correct number of container were running (and a couple seconds more) before running the tests.

```
# on both nodes
systemctl stop docker
rm -rf /var/lib/docker/*
systemctl start docker
# on d1
docker swarm init --listen-addr $ip1 --advertise-addr $ip1
# from here on everithing is run on d3
docker swarm join --listen-addr $ip3 --advertise-addr $ip3 --token $token $ip1
# docker service create --replicas 2 --name hostname --publish 80:80 alpine sh -c 'while true; do echo -e "HTTP/1.1 200 OK\n\n$(hostname)" | nc -lp 80; done'
dlkx7cxdrx4vv1e7zy7r71002
# test $ip1 |sort|uniq|wc -l;test $ip3|sort|uniq|wc -l
2
2
# docker service scale hostname=10
hostname scaled to 10
# test $ip1 |sort|uniq|wc -l;test $ip3|sort|uniq|wc -l
10
10
# docker service scale hostname=20
hostname scaled to 20
# test $ip1 |sort|uniq|wc -l;test $ip3|sort|uniq|wc -l
20
20
# docker service scale hostname=15
hostname scaled to 15
# test $ip1 |sort|uniq|wc -l;test $ip3|sort|uniq|wc -l
15
15
# docker service scale hostname=20
hostname scaled to 20
# test $ip1 |sort|uniq|wc -l;test $ip3|sort|uniq|wc -l
20
20
```

Up to here it run smootly, just as per your test.

But when I tried scaling the service back to 15:

```
# docker service scale hostname=15
hostname scaled to 15
# test $ip1 |sort|uniq|wc -l;test $ip3|sort|uniq|wc -l
12
15
```
In this first test requests arriving at $ip1 were routed to 12 containers, while requests arriving at $ip3 were routed to 15 containers.

When trying to dig deeper $ip3 started routing requests to 12 container too:
```
# test $ip1 |sort|uniq;echo;test $ip3|sort|uniq
2ea06ccf070b
2ec3133c10d9
49a2e11d1a85
62a6a054e6cc
7831c893a8e9
7d0b6f9d6bd7
904161490f98
aa5a81e69093
b378b205d88d
b3e9074c43e6
ddae1d803a72
e256d3030e81

2ea06ccf070b
2ec3133c10d9
49a2e11d1a85
62a6a054e6cc
7831c893a8e9
7d0b6f9d6bd7
904161490f98
aa5a81e69093
b378b205d88d
b3e9074c43e6
ddae1d803a72
e256d3030e81
# test $ip1 |sort|uniq;echo;test $ip3|sort|uniq
2ea06ccf070b
2ec3133c10d9
49a2e11d1a85
62a6a054e6cc
7831c893a8e9
7d0b6f9d6bd7
904161490f98
aa5a81e69093
b378b205d88d
b3e9074c43e6
ddae1d803a72
e256d3030e81

2ea06ccf070b
2ec3133c10d9
49a2e11d1a85
62a6a054e6cc
7831c893a8e9
7d0b6f9d6bd7
904161490f98
aa5a81e69093
b378b205d88d
b3e9074c43e6
ddae1d803a72
e256d3030e81
```

while the service has the correct number of tasks active:

```
# docker service ps hostname|grep Running
3q2vqb0ir9s4euoo529dhf8fq  hostname.1       alpine  d1    Running        Running 3 minutes ago        
5qxriec2xzzzt4hvppczwfj0p  hostname.5       alpine  d1    Running        Running 3 minutes ago        
31azn2lmapsmezgk67d7xsa8d  hostname.6       alpine  d3    Running        Running 2 minutes ago        
8ufc35sdvg3a2st8gp6rgvsan  hostname.7       alpine  d3    Running        Running 3 minutes ago        
3qft7skcnt30bc51u9fqbdova  hostname.9       alpine  d1    Running        Running 3 minutes ago        
2dg3o0lwfn2u7lwcauee9r9be  hostname.10      alpine  d3    Running        Running 3 minutes ago        
1by1ve6wuyi2jmodb5gqn8yqa  hostname.11      alpine  d3    Running        Running 2 minutes ago        
43gjp531w0zcs8o37j1xbkjha  hostname.12      alpine  d3    Running        Running 2 minutes ago        
39zpljojp6jg6040tbcomwvyk  hostname.13      alpine  d1    Running        Running 2 minutes ago        
0sav1mw0xwso5n0uvaj2udc4b  hostname.14      alpine  d1    Running        Running 2 minutes ago        
6vdgztb76jjk7yxy3qu9k5npy  hostname.15      alpine  d1    Running        Running 2 minutes ago        
4p79gc4g0881k6gzeod224j2a  hostname.17      alpine  d3    Running        Running 2 minutes ago        
5y2gycgko7e7yokon65hln5gj  hostname.18      alpine  d1    Running        Running 2 minutes ago        
0agz4dbfwfsc8dyd0ip509mrl  hostname.19      alpine  d1    Running        Running 2 minutes ago        
0m59sbvyckdfy8azqogjceq9n  hostname.20      alpine  d3    Running        Running 2 minutes ago        
# docker service ps hostname|grep Running|wc -l
15
```

After some 15 minutes the situation is unchanged
```
# test $ip1 |sort|uniq|wc -l;test $ip3|sort|uniq|wc -l
12
12
```

and subsequent scaling kept having a wacky number of container reached

```
# docker service scale hostname=17
hostname scaled to 17
# test $ip1 |sort|uniq|wc -l;test $ip3|sort|uniq|wc -l
14
14
# docker service scale hostname=14
hostname scaled to 14
# test $ip1 |sort|uniq|wc -l;test $ip3|sort|uniq|wc -l
11
11
```

Then I tried going back to scale=0 -> scale=4 , and everything was working
again.
But the problem presented itself again at the next scaling
```
# docker service scale hostname=0
hostname scaled to 0
# docker service scale hostname=4
hostname scaled to 4
# test $ip1 |sort|uniq|wc -l;test $ip3|sort|uniq|wc -l
4
4
# docker service scale hostname=10
hostname scaled to 10
root@d3:~# test $ip1 |sort|uniq|wc -l;test $ip3|sort|uniq|wc -l
7
7
```


The `test` function i used is defined as
```
function test() {  
  ip=$1
  for i in $(seq 1 50)
  do
    curl -s $ip
  done
}
```

**Comment by dariko** [comment]
I kept trying and after another scale=0, scale=10 the test got stuck on $ip3.
I tested it manually and noticed $ip3 had became unreachable on port 80, while dockerd was still a socket using it:
```
# docker service scale hostname=0
hostname scaled to 0
root@d3:~# docker service scale hostname=10
hostname scaled to 10
root@d3:~# test $ip1 |sort|uniq|wc -l;test $ip3|sort|uniq|wc -l
10
^C
#### here I ctrl-c'd the command after a minute or so
# curl $ip3
curl: (7) Failed to connect to 10.91.161.165 port 80: Connection refused
#### still on d3
# netstat -anp|grep STEN\ |grep 80
tcp6       0      0 :::80                   :::*                    LISTEN      23660/dockerd
```

And, while I was inspecting this situation,  I noticed the service had become unreachable even from d1, also with the port still in use by docker
```
# telnet $ip1 80
Trying 10.91.161.163...
telnet: Unable to connect to remote host: Connection refused
```

still the service has the correct number of tasks running
```
# docker service ps hostname|grep nning
3ycg8lfkrhjy4sjbnz16euv7w  hostname.1       alpine  d1    Running        Running 11 minutes ago   
5h9am522v7w76mm8f43tup9y3  hostname.2       alpine  d3    Running        Running 11 minutes ago   
5si1finnz6jzwctznixct0vqm  hostname.3       alpine  d1    Running        Running 11 minutes ago   
5sga89w3nayvz18d05eozmixo  hostname.4       alpine  d1    Running        Running 11 minutes ago   
0f69uiddfu2h8a744yyk0ty9w  hostname.5       alpine  d3    Running        Running 11 minutes ago   
6u4b3ah8lsdcgbwxqruoozr70  hostname.6       alpine  d3    Running        Running 11 minutes ago   
8s6p644ksrnny5px5wcli5g7n  hostname.7       alpine  d3    Running        Running 11 minutes ago   
138t8o6kqsqnw71kb9hz8qpq1  hostname.8       alpine  d1    Running        Running 11 minutes ago   
62024nfr0go3siuxotegwc43d  hostname.9       alpine  d1    Running        Running 11 minutes ago   
e8adh4z6zuw6hg8z7pzpfywd0  hostname.10      alpine  d3    Running        Running 11 minutes ago   
root@d1:~# docker service ps hostname|grep nning|wc -l
10
```

**Comment by sswierczyna** [comment]
I have similar issue after reboot one of node.
In `syslog` I can see:
```
Jul 30 21:57:13 swarm-a-02 dockerd[3623]: message repeated 2 times: [ time="2016-07-30T21:57:13+02:00" level=info msg="Firewalld running: false"]
Jul 30 21:57:39 swarm-a-02 dockerd[3623]: time="2016-07-30T21:57:39.554804332+02:00" level=error msg="could not resolve peer \"10.255.0.4\": could not resolve peer: serf instance not initialized"
Jul 30 21:58:08 swarm-a-02 kernel: [  249.361061] IPVS: rr: FWM 258 0x00000102 - no destination available
Jul 30 21:58:08 swarm-a-02 kernel: [  249.423575] IPVS: rr: FWM 258 0x00000102 - no destination available
Jul 30 21:58:08 swarm-a-02 kernel: [  249.483083] IPVS: rr: FWM 258 0x00000102 - no destination available

```

```
szop@swarm-a-02:~$ docker version 
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 22:11:10 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 22:11:10 2016
 OS/Arch:      linux/amd64
```

```
szop@swarm-a-02:~$ docker info
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 108
Server Version: 1.12.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: xfs
 Dirs: 62
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host null overlay
Swarm: active
 NodeID: bo9lhfnk7uiea6mxczo4n3fpa
 Is Manager: true
 ClusterID: 9s8f7x99fcmn8y17rjz6m1nhc
 Managers: 3
 Nodes: 4
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot interval: 10000
  Heartbeat tick: 1
  Election tick: 3
 Dispatcher:
  Heartbeat period: 5 seconds
 CA configuration:
  Expiry duration: 3 months
 Node Address: 10.0.8.171
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-31-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.796 GiB
Name: swarm-a-02
ID: 73Z2:PYEJ:X3WF:XSUY:G3YY:R7VW:UYAD:MND5:72ZZ:RXMN:GNC3:HEAG
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8

**Comment by dariko** [comment]
Resolved with 1.12.1
